OSDOCS-12920: oc adm upgrade recommend TP doc (#86227)
openshift/openshift-docs: modules/update-upgrading-oc-adm-upgrade-recommend.adoc

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/updating-cluster-cli.adoc

:_mod-docs-content-type: PROCEDURE
[id="update-upgrading-oc-adm-upgrade-recommend_{context}"]

> = Finding recommended update paths with `oc adm upgrade recommend` (Technology Preview)

Reviewer: It seem this style is a little different from `Gathering cluster update status using oc adm upgrade status (Technology Preview)`, could you please help take a look?

Author: Hi @JianLi-RH ! Happy to take a look. What specifically are you referring to?

Reviewer: hi @obrown1205, From the preview page, I can see `oc adm upgrade recommend` is shown as a code block, but `oc adm upgrade status` is not. And because I can't see the backend code, so need you confirm. Thanks!

Author: Ah! Fixed it :) Can you PTAL?

When updating your cluster, it is useful to find out which versions are recommended for your update. While the `oc adm upgrade` command returns a list of the next available versions, this command decouples that information and narrows down your suggestions to recommend a new target release before you launch your update.

The `oc adm upgrade recommend` command is read-only and will never alter any state in your cluster. 

:FeatureName: The `oc adm upgrade recommend` command
include::snippets/technology-preview.adoc[]

[IMPORTANT]
====

> While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `OC_ENABLE_CMD_UPGRADE_STATUS` Technology Preview environment variable, otherwise the {oc-first} will not recognize the command and you will not be able to use the feature. 

Reviewer: It seems forgot correct the env variable? `OC_ENABLE_CMD_UPGRADE_STATUS` should be `OC_ENABLE_CMD_UPGRADE_RECOMMEND`?

Author: good catch! TY!

====

.Procedure 

> . Set the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true` by running the following command:

Reviewer: Sorry, is it should be `environment variable` or `environmental variable`? sorry, I am not native speaker, I can't distinguish their difference.

+
[source,terminal]
----
$ export OC_ENABLE_CMD_UPGRADE_RECOMMEND=true
----
. Run the `oc adm upgrade recommend` command:
+
[source,terminal]
----
$ oc adm upgrade recommend
----
+

> .Example output for an update progressing successfully

Reviewer 2: If we have to name this output, I'd call it something like "Example output when 4.y is recommended unconditionally, while the longest 4.(y+1) have assessed issues", or something, but that's a bit of a mouthful.  I don't think we should use "for an update progressing successfully", because we're thinking about kicking off an update, and that's mostly happening while the cluster is not already mid-update.  Since we only have the one example, we can maybe just say "Example output"?

Author: That's fine by me :)

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/updating-cluster-cli.adoc

:_mod-docs-content-type: PROCEDURE
[id="update-upgrading-oc-adm-upgrade-recommend_{context}"]

> = Finding recommended update paths with `oc adm upgrade recommend` (Technology Preview)

Reviewer 3: No backticks in assembly or module headings. You can remove them, or think of another way to phrase this that doesn't use the command itself.
```suggestion
= Finding recommended update paths with oc adm upgrade recommend (Technology Preview)
```

Related to [this earlier comment](https://github.com/openshift/openshift-docs/pull/86227/files#r1893350155) I think

> When updating your cluster, it is useful to find out which versions are recommended for your update. While the `oc adm upgrade` command returns a list of the next available versions, this command decouples that information and narrows down your suggestions to recommend a new target release before you launch your update.

Reviewer 3: Try to avoid subjective value/judgement statements, also not sure we need to repeat update in the forst line:
```suggestion
You can find out which versions are recommended for your update. While the `oc adm upgrade` command returns a list of the next available versions, this command decouples that information and narrows down your suggestions to recommend a new target release before you launch your update.
```

"this command" is a bit unclear - I think referring to the new `recommend` command? So maybe clarify the antecedent:
```suggestion
You can use the `oc adm upgrade recommend` command to see which versions are recommended for your update. While the `oc adm upgrade` command returns a list of the next available versions, this command decouples that information and narrows down your suggestions to recommend a new target release before you launch your update.
```

It still has a somewhat unclear antecedent. What about flipping it around a bit?
```suggestion
When updating your cluster, the `oc adm upgrade` command returns a list of the next available versions. You can use the `oc adm upgrade recommend` command to narrow down your suggestions and recommend a new target release before you launch your update.
```

Author: great suggestion! TY!

> The `oc adm upgrade recommend` command is read-only and will never alter any state in your cluster. 

Reviewer 3: Does rephrasing to avoid future work ok here?
```suggestion
The `oc adm upgrade recommend` command is read-only and cannot alter the state of your cluster. 
```

:FeatureName: The `oc adm upgrade recommend` command
include::snippets/technology-preview.adoc[]

[IMPORTANT]
====
While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `OC_ENABLE_CMD_UPGRADE_RECOMMEND` Technology Preview environment variable, otherwise the {oc-first} will not recognize the command and you will not be able to use the feature. 

> ====

Reviewer 3: Any way around stacking two admonitions here?

Reviewer 3: after reading step 1: I would remove most of this and maybe just add it back as a note that a TP cluster isn't required, since you explicitly make them enable that flag in the proc anyway.

.Procedure 

> . Set the `OC_ENABLE_CMD_UPGRADE_RECOMMEND` environment variable to `true` by running the following command:

Reviewer 3: This is telling them to do the thing in the IMPORTANT box above. By virtue of it being a step in the proc, it's explicitly required.

+
[source,terminal]
----
$ export OC_ENABLE_CMD_UPGRADE_RECOMMEND=true
----
. Run one of the following commands:

> .. If you want a list of all of the recommended versions for your update, run the `oc adm upgrade recommend` command:

Reviewer 3: ```suggestion
.. To list the recommended versions for your update, run the `oc adm upgrade recommend` command:
```

Reviewer 3: Occurred to me after the fact that a and b should probably be `**` instead of `..` since IIUC they don't need to be run in order (or even need to both be run at all)

+
[source,terminal]
----
$ oc adm upgrade recommend
----
+

> .Example output
> [%collapsible]
> ====
> [source,terminal]
> ----
> Upstream is unset, so the cluster will use an appropriate default.
> Channel: stable-4.13 (available channels: candidate-4.12, candidate-4.13, eus-4.12, eus-4.14, fast-4.12, fast-4.13, stable-4.12, stable-4.13)
> 
> Updates to 4.13:
> 
> Version: 4.13.50
> Image: quay.io/openshift-release-dev/ocp-release@sha256:6afb11e1cac46fd26476ca134072937115256b9c6360f7a1cd1812992c065f02
> Reason: AdminAckRequired
> Message: Kubernetes 1.26 and therefore OpenShift 4.13 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958394 for details and instructions.
> 
> Version: 4.13.49
> Image: quay.io/openshift-release-dev/ocp-release@sha256:ab6f574665395d809511db9dc57764358278538eaae248c6d199208b3c30ab7d
> Reason: AdminAckRequired
> Message: Kubernetes 1.26 and therefore OpenShift 4.13 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958394 for details and instructions.
> 
> Updates to 4.12:
> VERSION     ISSUES
> 4.12.64     no known issues
> 4.12.63     no known issues
> And 43 older 4.12 updates you can see with '--show-outdated-releases' or '--version VERSION'.
> ----
> ====

Reviewer 3: Collapsible blocks don't work on docs.redhat.com :crying_cat_face: 

```suggestion
.Example output
[source,terminal]
----
Upstream is unset, so the cluster will use an appropriate default.
Channel: stable-4.13 (available channels: candidate-4.12, candidate-4.13, eus-4.12, eus-4.14, fast-4.12, fast-4.13, stable-4.12, stable-4.13)

Updates to 4.13:

Version: 4.13.50
Image: quay.io/openshift-release-dev/ocp-release@sha256:6afb11e1cac46fd26476ca134072937115256b9c6360f7a1cd1812992c065f02
Reason: AdminAckRequired
Message: Kubernetes 1.26 and therefore OpenShift 4.13 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958394 for details and instructions.

Version: 4.13.49
Image: quay.io/openshift-release-dev/ocp-release@sha256:ab6f574665395d809511db9dc57764358278538eaae248c6d199208b3c30ab7d
Reason: AdminAckRequired
Message: Kubernetes 1.26 and therefore OpenShift 4.13 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958394 for details and instructions.

Updates to 4.12:
VERSION     ISSUES
4.12.64     no known issues
4.12.63     no known issues
And 43 older 4.12 updates you can see with '--show-outdated-releases' or '--version VERSION'.
----
```

> .. If you want to determine if a specific version is recommended for your update, run the `oc adm upgrade recommend --version` command:

Reviewer 3: ```suggestion
.. To determine whether a specific version is recommended for your update, run the `oc adm upgrade recommend --version` command:
```

+
[source,terminal]
----
$ oc adm upgrade recommend --version 4.12.51
----
+

> .Example output
> [%collapsible]
> ====
> [source,terminal]
> ----
> Upstream is unset, so the cluster will use an appropriate default.
> Channel: stable-4.13 (available channels: candidate-4.12, candidate-4.13, eus-4.12, eus-4.14, fast-4.12, fast-4.13, stable-4.12, stable-4.13)
> 
> Update to 4.12.51 Recommended=False:
> Image: quay.io/openshift-release-dev/ocp-release@sha256:158ced797e49f6caf7862acccef58484be63b642fdd2f66e6416295fa7958ab0
> Release URL: https://access.redhat.com/errata/RHSA-2024:1052
> Reason: MultipleReasons
> Message: An unintended reversion to the default kubelet nodeStatusReportFrequency can cause significant load on the control plane. https://issues.redhat.com/browse/MCO-1094
>   
> After rebooting into kernel-4.18.0-372.88.1.el8_6 or later, kernel nodes experience high load average and io_wait times. The nodes might fail to start or stop pods and probes may fail. Workload and host processes may become unresponsive and workload may be disrupted. https://issues.redhat.com/browse/COS-2705
> ----
> ====

Reviewer 3: ```suggestion
.Example output
[source,terminal]
----
Upstream is unset, so the cluster will use an appropriate default.
Channel: stable-4.13 (available channels: candidate-4.12, candidate-4.13, eus-4.12, eus-4.14, fast-4.12, fast-4.13, stable-4.12, stable-4.13)

Update to 4.12.51 Recommended=False:
Image: quay.io/openshift-release-dev/ocp-release@sha256:158ced797e49f6caf7862acccef58484be63b642fdd2f66e6416295fa7958ab0
Release URL: https://access.redhat.com/errata/RHSA-2024:1052
Reason: MultipleReasons
Message: An unintended reversion to the default kubelet nodeStatusReportFrequency can cause significant load on the control plane. https://issues.redhat.com/browse/MCO-1094
  
After rebooting into kernel-4.18.0-372.88.1.el8_6 or later, kernel nodes experience high load average and io_wait times. The nodes might fail to start or stop pods and probes may fail. Workload and host processes may become unresponsive and workload may be disrupted. https://issues.redhat.com/browse/COS-2705
----
```

> With this information, you can make informed decisions on how to proceed with your update.

Reviewer 3: Shouldn't include plain text outside of steps in a `.Procedure` block. This is part of the user story for the proc, move it to the intro.
